internal/auth/authn/role_extractor.go

```diff
@@ -0,0 +1,122 @@
+package authn
+
+import (
+	"fmt"
+
+	api "github.com/flightctl/flightctl/api/v1alpha1"
+)
+
+// RoleExtractor handles role extraction from claims based on role assignment configuration
+type RoleExtractor struct {
+	roleAssignment api.AuthRoleAssignment
+}
+
+// NewRoleExtractor creates a new role extractor with the given role assignment
+func NewRoleExtractor(roleAssignment api.AuthRoleAssignment) *RoleExtractor {
+	return &RoleExtractor{
+		roleAssignment: roleAssignment,
+	}
+}
+
+// ExtractRolesFromMap extracts roles from a map of claims (for OAuth2 userinfo)
+func (r *RoleExtractor) ExtractRolesFromMap(claims map[string]interface{}) []string {
+	discriminator, err := r.roleAssignment.Discriminator()
+	if err != nil {
+		return []string{}
```

Comment:
minor: go treats nil and empty slices the same
```suggestion
		return nil
```

Reply:
adjusted